### ISSUE TITLE
NLL migration mode sometimes prints more diagnostics than either AST-borrowck or NLL

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#53004]
While investigating #52979, I noticed some of the diagnostics from `--edition 2018` are different than what either of the `test.stderr` and `test.nll.stderr` would lead one to expect to see.

This appears to arise from the implementation of `-Z borrowck=migrate`, from what I can tell.

For example, on ui/borrowck/borrowck-in-static.rs, we have from AST-borrowck:

https://github.com/rust-lang/rust/blob/40cb4478a3f550bf12d81cdcf08cc9ef3985ed41/src/test/ui/borrowck/borrowck-in-static.stderr#L1-L9

and from NLL we have:

https://github.com/rust-lang/rust/blob/40cb4478a3f550bf12d81cdcf08cc9ef3985ed41/src/test/ui/borrowck/borrowck-in-static.nll.stderr#L1-L7

but I am seeing *this* from `-Z borrowck=migrate`:

```
error[E0507]: cannot move out of captured variable in an `Fn` closure
  --> ../src/test/ui/borrowck/borrowck-in-static.rs:15:17
   |
15 |     Box::new(|| x) //~ ERROR cannot move out of captured outer variable
   |                 ^ cannot move out of captured variable in an `Fn` closure

error[E0507]: cannot move out of `x`, as it is a captured variable in a `Fn` closure
  --> ../src/test/ui/borrowck/borrowck-in-static.rs:15:17
   |
15 |     Box::new(|| x) //~ ERROR cannot move out of captured outer variable
   |                 ^
   |                 |
   |                 cannot move out of `x`, as it is a captured variable in a `Fn` closure
   |                 cannot move
   |
help: consider changing this to accept closures that implement `FnMut`
  --> ../src/test/ui/borrowck/borrowck-in-static.rs:15:14
   |
15 |     Box::new(|| x) //~ ERROR cannot move out of captured outer variable
   |              ^^^^

error: aborting due to 2 previous errors
```

My guess that this is arising because I must have missed at least one case where AST-borrowck signals an error. Thus, when the `-Z borrowck=migrate` runs the AST-borrowck as a fallback in response to an error encounters during the MIR-borrowck, the user can sometimes see the *both sets* the errors from both NLL and AST-borrowck.

**Comment by pnkfelix** [comment]
Actually my hypothesis that this is the sum of the two sets of errors is clearly false given even just the simple example above. Where does `-Z borrowck=migrate` get the message:

```
error[E0507]: cannot move out of `x`, as it is a captured variable in a `Fn` closure
```

when that particular level of detail is not included in either of the other diagnostics presented in the other two files...?

**Comment by pnkfelix** [comment]
Ah wait, this was actually an expected side-effect of migrate mode, due to this part of #52681

https://github.com/rust-lang/rust/blob/4dae47051315793d4d4d043b4b8b03c51e3dc877/src/librustc_mir/borrow_check/mod.rs#L1920-L1958

**Comment by pnkfelix** [comment]
So there a some follow up questions I now have:

 1. I do remember going through the control-flow for this case and figuring out why it "made sense" that this code path became reachable under migration mode. But I have since then forgotten the justification. It would be good to at least document the justification in a comment.
  2. More upsetting (to me at least): the new error we get here under migration mode, presumably generated via `self.report_mutability_error(..)`, looks slightly better to me than the errors we are getting from AST-borrowck and NLL: it names `x`, and (more importantly) it provides a *hint* that you may want to change the code so that it accepts `FnMut` closures...

**Comment by pnkfelix** [comment]
Having said that, neither of these follow up questions is sufficiently high priority for me to spend more time on them today. I'm going to unassign myself from this issue, but leave it open so that people can find it if they similarly encounter this arguably bizarre behavior from migration mode (which is turned on in `--edition 2018` ...)

**Comment by pnkfelix** [comment]
although now that I think about it more ... I'm *certain* that in my original justification for the newly generated diagnostics from cases like `borrowck-in-static.rs`, I had thought it could only arise in cases where the NLL errors would be downgraded to warnings; i.e. situations where AST-borrowck did not error.

but clearly this is a case where AST-borrowck *did* error, or at least something in that pipeline errored.

So something is still funky here.

**Comment by nikomatsakis** [comment]
@pnkfelix I am doing a triage sweep and am not sure where to put this -- going to put in Edition Release for now I guess.

**Comment by pnkfelix** [comment]
A lot of the cases here have been addressed, especially by PR #55221.

The 3rd NLL diagnostic review (#54528) contains an updated list of all the diagnostic deviations.

I suspect at this point that (and/or the [associated project](https://github.com/rust-lang/rust/projects/10)) should be the canonical database for listing deviations like the ones listed here.

So, I'm closing this issue.
